scripts/detect_duplicate_features.py

```diff
@@ -0,0 +1,89 @@
+import os
+import argparse
+
+import yaml
+
+
+def findall_features(features):
+    feature_list = []
+    for feature in features:
+        if "and" in feature:
+            and_list = findall_features(feature["and"])
+            for x in and_list:
+                feature_list.append(x)
+        elif "or" in feature:
+            or_list = findall_features(feature["or"])
+            for y in or_list:
+                feature_list.append(y)
+        else:
+            feature_list.append(feature)
+
+    return feature_list
+
+
+def find_overlapping_rules(new_rule_path, rules_path):
+    if not new_rule_path.endswith(".yml"):
+        return "ERROR ! New rule path file name incorrect"
+
+    count = 0
+
+    with open(new_rule_path, "r") as f:
+        new_rule = yaml.safe_load(f)
+
+    if "rule" not in new_rule:
+        return "ERROR ! given new rule path isn't a rule"
+
+    new_rule_features = findall_features(new_rule["rule"]["features"])
+
+    overlapping_rules = []
+
+    for dirpath, dirnames, filenames in os.walk(rules_path):
+        for filename in filenames:
+            if filename.endswith(".yml"):
+                rule_path = os.path.join(dirpath, filename)
+                with open(rule_path, "r") as f:
+                    rule = yaml.safe_load(f)
+                    if "rule" not in rule:
+                        continue
+                    rule_features = findall_features(rule["rule"]["features"])
+                    count += 1
+                if any([feature in rule_features for feature in new_rule_features]):
+                    overlapping_rules.append(rule_path)
+
+    result = {"overlapping_rules": overlapping_rules, "count": count}
+
+    return result
+
+
+# python script.py --base-dir /path/to/capa/rules rules/anti-analysis/reference-analysis-tools-strings.yml rules
+
+
+def main():
+    # usage
+
+    parser = argparse.ArgumentParser(description="Find overlapping rules in Capa rules.")
+    parser.add_argument("-b", "--base-dir", default="", help="Base directory for Capa rules.")
+    parser.add_argument("-f", "--new_rule_path", required=True, help="Path to the new Capa rule.")
+    parser.add_argument("-d", "--rules_path", required=True, help="Path to the directory containing Capa rules.")
+    args = parser.parse_args()
+
+    base_dir = args.base_dir
+    new_rule_path = os.path.join(base_dir, args.new_rule_path)
+    rules_path = os.path.join(base_dir, args.rules_path)
+
+    try:
+        result = find_overlapping_rules(new_rule_path, rules_path)
+        print("New rule path : %s" % new_rule_path)
+        print("Number of rules checked : %s " % result["count"])
+        print("Paths to overlapping rules : ", result["overlapping_rules"])
+        print("Number of rules containing same features : %s" % len(result["overlapping_rules"]))
+    except Exception as e:
```

Comment:
does this mark any unexpected issues? what exceptions do you expect here?

Reply:
It will show if any Exceptions raised in find_overlapping_features function.

Reply:
Can you list them explicitly for Exceptions you'd expect here?

Reply:
There can 2 exceptions.
- Path to new rules isn't a yaml file
- The new rules files isn't a files containing any rules. There can be yaml files whose top level element is not a rule.

Reply:
lets' catch these explicitly instead of all Exceptions

Reply:
> lets' catch these explicitly instead of all Exceptions

I don't think there is a need for that.
If there are any exceptions it prints them as - files which faced error + Error class + Error.

Reply:
Which could mask other errors in the program, see https://www.flake8rules.com/rules/E722.html
I'm not super concerned about it here, but as a general concept.